File: C++/261. Graph Valid Tree.cpp
class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        int m = edges.size();
        if(m != n - 1) return false;

        vector<vector<int>> g = vector<vector<int>>(n, vector<int>());
        for(auto &v : edges) {
            auto &a = v.front(), &b = v.back();
            g[b].push_back(a);
            g[a].push_back(b);
        }

        unordered_set<int> seen;
        auto dfs = [&](this auto &&dfs, int root) -> void {
            seen.insert(root);
            for(auto &e : g[root]) {
                if(seen.contains(e)) continue;
                dfs(e);
            }
        };
        dfs(0);
        return seen.size() == n;
    }
};

File: Swift/261 Graph Valid Tree.swift
class Solution {
    func validTree(_ n: Int, _ edges: [[Int]]) -> Bool {
        if n != edges.count + 1 {
            return false
        }
        var g = [[Int]](repeating: [], count: n)
        for item in edges {
            if let a = item.first, let b = item.last {
                g[a].append(b)
                g[b].append(a)
            }
        }

        var seen = Set<Int>()
        func dfs(_ root: Int) {
            seen.insert(root)
            for item in g[root] {
                if !seen.contains(item) {
                    dfs(item)
                }
            }
        }
        dfs(0)
        return seen.count == n
    }
}
